### PR TITLE
refactor: Update to use EKS module which follows EKS Blueprints v5 direction

### PR DESCRIPTION
**Pull request by bryantbiggs**
#### What this PR does / why we need it:
- Update to use EKS module which follows EKS Blueprints v5 direction

#### Which issue(s) this PR fixes:

Reference https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/1421

#### Quality checks

- [ ] My content adheres to the style guidelines
- [ ] I ran `make test` or `make e2e-test` and it was successful

I tried to run `make test` but got this:
```
make test
bash hack/run-tests.sh 'terraform' '*'
Error: Please specify a module
make: *** [test] Error 1
```

And when I try `make e2e-test`:
```
make e2e-test
make: *** No rule to make target `e2e-test'.  Stop.
```

I don't see an `e2e-test` target in the Makefile, and I'm not sure what a "module" is to target with `make test`. Any pointers on how to test would be appreciated 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: terraform/modules/cluster/addons.tf [hunk 1]
```diff
@@ -1,35 +1,17 @@
-data "aws_eks_addon_version" "latest" {
-  for_each = toset(["vpc-cni"])
-
-  addon_name         = each.value
-  kubernetes_version = var.cluster_version
-  most_recent        = true
-}
-
-resource "aws_eks_addon" "vpc_cni" {
```
Comment (bryantbiggs) [inline terraform/modules/cluster/addons.tf:9]: Based on @niallthomson's findings, the EKS module was updated to allow configuring addons before compute is ready https://github.com/terraform-aws-modules/terraform-aws-eks/pull/2478

In addition, there is now a variable `dataplane_wait_duration` within the module that users can set a variable time duration between the control plane becoming active, and the start of downstream compute resource provisioning. Currently it defaults to `"30s"`

File: terraform/modules/cluster/addons.tf [hunk 2]
```diff
@@ -1,35 +1,17 @@
...
-  depends_on = [
-    null_resource.kubectl_set_env
-  ]
-}
-
 locals {
   ebs_csi_blocker = try(module.eks_blueprints_kubernetes_addons.aws_ebs_csi_driver.release_metadata.metadata.status, "")
 }
 
+resource "aws_iam_instance_profile" "karpenter" {
```
Comment (bryantbiggs) [inline terraform/modules/cluster/addons.tf:5]: once https://github.com/aws-ia/terraform-aws-eks-blueprints-addons is ready, this will go away. This [sub-module](https://github.com/terraform-aws-modules/terraform-aws-eks/tree/master/modules/karpenter) provisions all of the necessary AWS infrastructure resources, including this profile. The new addons module will use this plus the helm release for karpenter which will alleviate the need for this resource

File: terraform/modules/cluster/addons.tf [hunk 3]
```diff
@@ -1,35 +1,17 @@
...
+
 module "eks_blueprints_kubernetes_addons" {
   source = "github.com/aws-ia/terraform-aws-eks-blueprints?ref=v4.16.0//modules/kubernetes-addons"
 
-  depends_on = [
-    aws_eks_addon.vpc_cni
-  ]
-
-  eks_cluster_id = module.eks_blueprints.eks_cluster_id
+  eks_cluster_id = module.eks.cluster_name
```
Comment (bryantbiggs) [inline terraform/modules/cluster/addons.tf:14]: You will see a lot of `*cluster_id` references being replaced with `*cluster_name`. You can see more details about this change here https://github.com/hashicorp/terraform-provider-aws/issues/27560

File: terraform/modules/cluster/addons.tf [hunk 4]
```diff
@@ -431,14 +413,10 @@ module "eks_blueprints_kubernetes_addons" {
 module "eks_blueprints_ack_addons" {
   source = "github.com/aws-ia/terraform-aws-eks-ack-addons?ref=v1.1.0"
 
-  depends_on = [
```
Comment (bryantbiggs) [inline terraform/modules/cluster/addons.tf:434]: the module waits for the nodegroups, which they wait for the VPC CNI due to the `before_compute` which means this explicit dependency can be removed

File: terraform/modules/cluster/addons.tf [hunk 5]
```diff
@@ -474,17 +452,15 @@ resource "random_string" "fluentbit_log_group" {
 }
 
 locals {
-  oidc_url = replace(data.aws_eks_cluster.cluster.identity[0].oidc[0].issuer, "https://", "")
```
Comment (bryantbiggs) [inline terraform/modules/cluster/addons.tf:477]: the EKS module outputs an attribute [`oidc_provider`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/df1b62548c1d8c7117f4ab45c5b494de64b34cb8/outputs.tf#L106) which is the issuer url with the `https://` stripped. It is named `oidc_provider` because this is the name assigned in the console, regardless of whether the provider was created by EKS or not

[`cluster_oidc_issuer_url`](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/df1b62548c1d8c7117f4ab45c5b494de64b34cb8/outputs.tf#L30) is the form with the fully qualified URL including `https://`

tl;dr - OIDC provider is whats used for IRSA, not the issuer URL

File: terraform/modules/cluster/eks.tf [hunk 6]
```diff
@@ -4,20 +4,38 @@ locals {
...
+
+          POD_SECURITY_GROUP_ENFORCING_MODE = "standard"
+        }
+      })
+    }
+  }
 
-  cluster_name    = var.environment_name
-  cluster_version = var.cluster_version
+  kms_key_enable_default_policy = true
```
Comment (bryantbiggs) [inline terraform/modules/cluster/eks.tf:35]: replicates prior functionality to ensure no KMS key conflicts https://github.com/terraform-aws-modules/terraform-aws-kms/blob/87be9ccb61f073b58ed56de0456214c34d793841/main.tf#L89-L103

The KMS key created by the EKS module makes the cluster creator identity the key admin

File: terraform/modules/cluster/eks.tf [hunk 7]
```diff
@@ -44,122 +58,77 @@ module "eks_blueprints" {
       type        = "ingress"
       self        = true
     }
-    # Recommended outbound traffic for Node groups
-    egress_all = {
```
Comment (bryantbiggs) [inline terraform/modules/cluster/eks.tf:48]: These rules are now added by default with the EKS module when `node_security_group_enable_recommended_rules` is `true` (which is the default) - https://github.com/terraform-aws-modules/terraform-aws-eks/blob/df1b62548c1d8c7117f4ab45c5b494de64b34cb8/node_groups.tf#L129-L180

The default rules do permit 1025-65535 for cross security group communication, but I suspect that with the sample applications there are additional ports (i.e. 80) required. I wasn't sure what was used so I left the custom rule addition for all cross security group communication open

File: terraform/modules/cluster/eks.tf [hunk 8]
```diff
@@ -44,122 +58,77 @@ module "eks_blueprints" {
...
-      instance_types  = ["m5.large"]
-      subnet_ids      = local.private_subnet_ids
-      min_size        = local.default_mng_min
-      max_size        = local.default_mng_max
-      desired_size    = local.default_mng_size
+  eks_managed_node_groups = {
+    managed-ondemand = {
+      instance_types = ["m5.large"]
 
-      ami_type        = "AL2_x86_64"
```
Comment (bryantbiggs) [inline terraform/modules/cluster/eks.tf:103]: The way EKS Blueprints managed this is a bit different from the EKS module. By default the `ami_type` is `AL2_x86_64`, and the release version is selected at the time of creation. I'm not sure what the explicit release version is used for in this context, typically we pull the latest version which includes the latest patches/etc. at the time of nodegroup creation. Let me know if this release version need to be put back

Ref: https://github.com/aws/containers-roadmap/issues/1945

File: terraform/modules/cluster/eks.tf [hunk 9]
```diff
@@ -44,122 +58,77 @@ module "eks_blueprints" {
...
+    managed-system = {
+      instance_types = ["m5.large"]
 
-      ami_type        = "AL2_x86_64"
-      release_version = var.ami_release_version
+      min_size     = 1
+      max_size     = 2
+      desired_size = 1
 
-      k8s_taints = [{ key = "systemComponent", value = "true", effect = "NO_SCHEDULE" }]
```
Comment (bryantbiggs) [inline terraform/modules/cluster/eks.tf:123]: I did have one question - this nodegroup, "managed-system" has a taint, and the nodegroup directly below, "managed-ondemand-tainted", does not despite its name containing `taint`. Should the taint be on that nodegroup?

File: terraform/modules/cluster/eks.tf [hunk 10]
```diff
@@ -177,54 +146,5 @@ resource "aws_security_group_rule" "dns_tcp" {
...
-  provisioner "local-exec" {
-    interpreter = ["/bin/bash", "-c"]
-    environment = {
-      KUBECONFIG = base64encode(local.kubeconfig)
-    }
-
-    # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
-    command = <<-EOT
-      sleep 30
-      kubectl set env daemonset aws-node -n kube-system POD_SECURITY_GROUP_ENFORCING_MODE=standard --kubeconfig <(echo $KUBECONFIG | base64 --decode)
```
Comment (bryantbiggs) [inline terraform/modules/cluster/eks.tf:226]: added into the VPC CNI custom configuration so all of this can be removed. Let me know if there is a reason this needs to be done this way instead of the custom config (currently both routes are shown/used)

**Comment by niallthomson** [comment]
Going to close this since the rearchitecture work will move away from this design
